generators/typescript/codegen/src/ast/core/Writer.ts

```diff
@@ -1,7 +1,149 @@
+            }
+            const moduleForAlias = this.starImportAliasesInverse[reference.importFrom.starImportAlias];
+            if (moduleForAlias != null && moduleForAlias !== reference.importFrom.moduleName) {
+                throw new Error(
+                    `Attempted to use alias ${reference.importFrom.starImportAlias} for more than one ` +
+                        "module in the same file"
+                );
+            }
+            const existingAlias = this.starImportAliases[reference.importFrom.moduleName];
+            if (existingAlias == null) {
+                this.starImportAliases[reference.importFrom.moduleName] = reference.importFrom.starImportAlias;
+                this.starImportAliasesInverse[reference.importFrom.starImportAlias] = reference.importFrom.moduleName;
+            } else if (existingAlias != null && existingAlias !== reference.importFrom.starImportAlias) {
+                throw new Error(
+                    "Cannot have more than one alias for non-named imports from a module: " +
+                        `got ${reference.importFrom.starImportAlias} but already have ${existingAlias}.`
+                );
+            }
+        }
+    }
+
     public toString(): string {
-        return this.buffer;
+        return this.stringifyImports() + this.buffer;
+    }
+
+    private stringifyImports(): string {
+        let result = "";
+        for (const [module, references] of Object.entries(this.imports)) {
+            const defaultImport = this.defaultImports[module];
+            let stringifiedNonDefault = "";
+            const named = references.filter((r) => r.importFrom?.type === "named");
+            const starImportAlias = this.starImportAliases[module];
+            if (named.length > 0 || defaultImport != null || starImportAlias != null) {
+                result += "import";
+                if (defaultImport != null) {
+                    result += ` ${defaultImport.name}`;
+                }
+                if (named.length > 0) {
+                    for (const ref of named.slice(0, -1)) {
+                        stringifiedNonDefault += `${ref.name}, `;
+                    }
+                    const lastRef = named[named.length - 1];
+                    // Need for eslint; lastRef will not be null because length > 0
+                    if (lastRef != null) {
+                        stringifiedNonDefault += `${lastRef.name}`;
+                    }
+                    if (defaultImport != null) {
+                        result += ",";
+                    }
+                    result += ` { ${stringifiedNonDefault} }`;
+                }
+                if (starImportAlias != null) {
+                    if (defaultImport != null || named.length > 0) {
+                        result += ", ";
+                    }
+                    result += ` * as ${starImportAlias}`;
+                }
+                result += ` from "${module}";\n`;
+            }
```

Comment:
We might be able to reduce nesting here too if we rearrange things a bit.

Reply:
What about something like this? We basically have a function that can turn a single module + multiple imports into statements. each statement is just an import

```ts
type Import = DefaultImport | NamedImport | StartImport; 

export function stringifyImports(): string[] {
  const importsByModule: Record<Module, Import[]>  = getImportsByModule();
  const statements = Object.entries(importsByModule).flatMap((module, imports) => {
     return  generateStatementsForImports(module, imports); 
  });
  return statements;
}

export function generateStatementsForImports(module, imports): string[] {
  const statements = []
  const nonStartImportStatement: string = "";
  for (const import of imports) {
    switch (import.type) {
       case "named": 
          nonStartImportStatement.push(named.value);
        case "startImport": 
           statements.push(`import * as ${import.alias} from ${module}`)
    }
  }
  return statements;
}
```